### PR TITLE
sriov, map pci-addresses to intefaces with network-status

### PR DESCRIPTION
**Pull request by RamLavi**
**What this PR does / why we need it**:
Currently the pci-addresses are mapped to the SRIOV interfaces using theSRIOV resource pool [0].
This is not enough to connect when there is more than 1 interfaces from the same resource pool, or when the interfaces use the same network-attachement-definition.
    
This commit utilies the network-status annotation from the vmi to accurately map the pci-address to each SRIOV interface, while keeping the legacy mapping method in case the annotation does not exist.

[0] https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin#configurations

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by orelmisan — COMMENTED** [review]
Thank you for the PR @RamLavi.
Please see the inline comments.

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool.go [hunk 1]
```diff
@@ -20,35 +20,67 @@
...
 
 // NewPCIAddressPool creates a PCI address pool based on the provided list of interfaces and
 // the environment variables that describe the SRIOV devices.
-func NewPCIAddressPool(ifaces []v1.Interface) *PCIAddressPool {
+func NewPCIAddressPool(sriovIfaces []v1.Interface, multusNetworks []v1.Network, vmiAnnotations map[string]string) *PCIAddressPool {
 	pool := &PCIAddressPool{
-		networkToResource: make(map[string]string),
+		networkToResource:         make(map[string]string),
+		interfaceNameToPciAddress: make(map[string]string),
+		legacyMode:                false,
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool.go:51]: Done

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool.go [hunk 2]
```diff
@@ -20,35 +20,67 @@
...
 	}
-	pool.loadResourcesNames(ifaces)
-	pool.loadResourcesAddresses()
+
+	err := pool.loadInterfaceNameToPciAddress(sriovIfaces, multusNetworks, vmiAnnotations)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool.go:55]: Done

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool.go [hunk 3]
```diff
@@ -20,35 +20,67 @@
...
 }
 
-func (p *PCIAddressPool) loadResourcesNames(ifaces []v1.Interface) {
-	for _, iface := range ifaces {
+func (p *PCIAddressPool) loadInterfaceNameToPciAddress(sriovIfaces []v1.Interface, multusNetworks []v1.Network, vmiAnnotations map[string]string) error {
+	sriovInterfaceToMultusInterfaceNameMap := loadInterfaceToMultusInterfaceName(sriovIfaces, multusNetworks)
+	multusInterfaceNameToNetworkStatusMap, err := loadMultusInterfaceNameToNetworkStatus(vmiAnnotations)
+	if err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool.go:69]: Agree. Done

**Comment by RamLavi** [comment]
> Thank you for the PR @RamLavi. Please see the inline comments.

fixed [Here](https://github.com/kubevirt/kubevirt/compare/eab656c81f0936249fa6d4f0eca48854003b4270..6035bf821e7f4fd2feac9593e03d90fbeb0994e6)

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	return nil
+}
+
+func getNetworkStatusList(vmiAnnotations map[string]string) ([]networkv1.NetworkStatus, error) {
+	netStatusesJson, isSet := vmiAnnotations[networkv1.NetworkStatusAnnot]
+	if !isSet {
+		return nil, errors.New(fmt.Sprintf("%s annotation is not present on vmi annotations", networkv1.NetworkStatusAnnot))
+	}
+
+	var netStatuses []networkv1.NetworkStatus
```
Comment (orelmisan) [inline pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go:123]: nit: Please consider this change so the returned value's name will match to the function's name:
```suggestion
	var networkStatusList []networkv1.NetworkStatus
```

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+func loadNetworkNameToMultusInterfaceName(interfaceList []v1.Interface, networkList []v1.Network) map[string]string {
+	networkToMultusInterfaceNameMap := map[string]string{}
+	for i, network := range networkList {
+		// multusInterfaceName follows a predefined indexation of the non-default multus networks (See pkg/virt-controller/services/multus_annotations.go: generateMultusCNIAnnotation.
+		// In order to not break this convention, this for loop should not break/continue
+		multusInterfaceName := fmt.Sprintf("net%d", i+1)
+		associatedNetwork := getIfaceByNetworkName(interfaceList, network.Name)
+		if associatedNetwork != nil {
+			networkToMultusInterfaceNameMap[network.Name] = multusInterfaceName
+		}
```
Comment (orelmisan) [inline pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go:90]: It is not clear from the variables names that this logic only maps the network names to which SR-IOV interfaces are connected.

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+		p.networkNameToPciAddress[networkName] = pciAddress
+	}
+	return nil
+}
+
+func (p *PCIAddressPoolWithNetworkStatus) Pop(networkName string) (string, error) {
+	pciAddress, exists := p.networkNameToPciAddress[networkName]
+	if !exists {
+		return "", fmt.Errorf("resource for SR-IOV network %s does not exist", networkName)
```
Comment (orelmisan) [inline pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go:74]: A resource has a different meaning in our context.

```suggestion
		return "", fmt.Errorf("PCI address for SR-IOV network %s does not exist", networkName)
```

**Comment by RamLavi** [comment]
[added](https://github.com/kubevirt/kubevirt/compare/158de568d77ffd643f5b1211d3526403d92fba62..c290cb43394b3210c67734d3a821c00e30619d87) e2e test that corresponds with the case the customer is reporting

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+	}
+
+	if err := pool.loadNetworkNameToPciAddress(sriovIfaces, multusNetworks, vmiAnnotations); err != nil {
+		return nil, err
+	}
+
+	return pool, nil
+}
+
+func (p *PCIAddressPoolWithNetworkStatus) loadNetworkNameToPciAddress(sriovIfaces []v1.Interface, multusNetworks []v1.Network, vmiAnnotations map[string]string) error {
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go:49]: Done

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+	pciAddress, exists := p.networkNameToPciAddress[networkName]
+	if !exists {
+		return "", fmt.Errorf("resource for SR-IOV network %s does not exist", networkName)
+	}
+
+	delete(p.networkNameToPciAddress, networkName)
+	return pciAddress, nil
+}
+
+func loadNetworkNameToMultusInterfaceName(interfaceList []v1.Interface, networkList []v1.Network) map[string]string {
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go:81]: Done

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+		multusInterfaceName := fmt.Sprintf("net%d", i+1)
+		associatedNetwork := getIfaceByNetworkName(interfaceList, network.Name)
+		if associatedNetwork != nil {
+			networkToMultusInterfaceNameMap[network.Name] = multusInterfaceName
+		}
+	}
+	return networkToMultusInterfaceNameMap
+}
+
+func loadMultusInterfaceNameToNetworkStatus(vmiAnnotations map[string]string) (map[string]networkv1.NetworkStatus, error) {
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go:95]: Done

**Comment by RamLavi** [comment]
This PR is no longer considered for merge. Closing

**Comment by moshe010** [comment]
@RamLavi  can you elaborate why This PR is no longer considered for merge? 
Are you working on alternative path?

**Comment by RamLavi** [comment]
> @RamLavi can you elaborate why This PR is no longer considered for merge? Are you working on alternative path?

yes of course. TL;DR, we are now considering mapping the sriov networks to PCI on the controller and using DownwardAPI to digest it on virt-launcher. (POC PR is in the making)
This option will be further discussed on the kubevirt community talk once the design PR is published.

Long story is, this PR is using the network-status annotation copied from the virt-launcher pod to the vmi instance in order to map the SRIOV networks to the PCI-Addresses. This approach is less straight forward when considering that during migration 2 virt-launcher pods are "owned" the same vmi, which complicates things. 
I think that the best approach is to use DownwardAPI of the SRIOV mapping so that it could be digested on virt-launcher, but of course this should and will be further discussed once the design is published.

**Comment by moshe010** [comment]
Do you have design doc that capture all of this or you first working on POC with downward api and after that you will have design doc?

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go (hunk 4 above)
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go:123]: Done

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go (hunk 5 above)
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go:90]: Fixed

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go (hunk 6 above)
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go:74]: I agree. Done

**Comment by RamLavi** [comment]
> Do you have design doc that capture all of this or you first working on POC with downward api and after that you will have design doc?

There is already a design doc, that was reviewed internally inside the team and will soon be published in a PR.
This design introduces the problem, and proposes the solution, and also other alternatives and explenations why we don't think we should pursue them. 
The POC PR is done in parallel, and will represent the solution we think is best. I will publish it here for once it is pushed

**Pull request closed without merging**